### PR TITLE
Organize intrinsics promotion checks

### PR DESCRIPTION
**Pull request by oli-obk**
cc @vertexclique 

supersedes #61835

r? @RalfJung

### CODE REVIEW

File: src/librustc_mir/transform/qualify_consts.rs [hunk 1]
```diff
@@ -660,6 +659,82 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
...
+                .opts
+                .debugging_opts
+                .unleash_the_miri_inside_of_you;
+            debug!(
+                "check_fn_in_const_ctx: {:?} is const fn: {:?}",
+                def_id, self.tcx.is_const_fn(def_id),
+            );
+            debug!("is_raw_const_fn: {}", self.tcx.is_const_fn_raw(def_id));
+            debug!("is_unstable: {:?}", self.tcx.is_unstable_const_fn(def_id));
+            debug!("stability: {:?}", self.tcx.lookup_stability(def_id));
```
Comment (spastorino) [inline src/librustc_mir/transform/qualify_consts.rs]: @oli-obk you probably want to remove this.

Comment (vertexclique) [reply src/librustc_mir/transform/qualify_consts.rs]: That was for debugging purposes I think, it will be removed when we are about to merge. Right? @oli-obk

Comment (oli-obk) [reply src/librustc_mir/transform/qualify_consts.rs]: Yes

File: src/libcore/intrinsics.rs [hunk 2]
```diff
@@ -940,6 +940,7 @@ extern "rust-intrinsic" {
     /// }
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
+    #[rustc_const_unstable(feature = "const_transmute")]
     pub fn transmute<T, U>(e: T) -> U;
```
Comment (RalfJung) [inline src/libcore/intrinsics.rs]: So we don't use `const fn` to mark intrinsics as const-evaluable, but we do use an attribute here to mark its const-nature as unstable? That is somewhat inconsistent I have to say.

File: src/librustc/ty/constness.rs [hunk 3]
```diff
@@ -63,13 +64,63 @@ impl<'tcx> TyCtxt<'tcx> {
...
+                    | sym::simd_extract
+                    | sym::simd_insert
+                    | sym::caller_location
+                    | sym::type_name
+                    => true,
+
+                    _ => false
+                }
+            }
+            _ => false
```
Comment (RalfJung) [inline src/librustc/ty/constness.rs]: Shouldn't this `bug!` as it is not an intrinsic?

File: src/librustc/ty/constness.rs [hunk 4]
```diff
@@ -63,13 +64,63 @@ impl<'tcx> TyCtxt<'tcx> {
...
+                    | sym::bswap
+                    | sym::bitreverse
+                    | sym::ctpop
+                    | sym::cttz
+                    | sym::cttz_nonzero
+                    | sym::ctlz
+                    | sym::ctlz_nonzero
+                    | sym::overflowing_add
+                    | sym::overflowing_sub
+                    | sym::overflowing_mul
```
Comment (RalfJung) [inline src/librustc/ty/constness.rs]: None of these seem to be implemented in `interpret/intrinsics.rs`?

File: src/librustc/ty/constness.rs [hunk 5]
```diff
@@ -63,13 +64,63 @@ impl<'tcx> TyCtxt<'tcx> {
...
 
 pub fn provide(providers: &mut Providers<'_>) {
-    /// only checks whether the function has a `const` modifier
+    /// Const evaluability whitelist is here to check evaluability at the
+    /// top level beforehand.
+    fn is_const_intrinsic(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
+        match tcx.fn_sig(def_id).abi() {
+            Abi::RustIntrinsic |
+            Abi::PlatformIntrinsic => {
+                match tcx.item_name(def_id) {
```
Comment (RalfJung) [inline src/librustc/ty/constness.rs]: `pref_align_of` and `ptr_offset_from` are missing from this list.

Could you add comments here and in `interpret/intrinsics.rs` saying that these lists need to be kept in sync? Also it would help if you could sort this list the same way  `interpret/intrinsics.rs` is sorted, to make it easier to match them up.

File: src/librustc_mir/transform/qualify_consts.rs [hunk 6]
```diff
@@ -660,6 +659,82 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
         }
     }
 
+    /// Check fn in const context with supplied feature gates
+    fn check_fn_in_const_ctx(&mut self, def_id: DefId) -> () {
```
Comment (RalfJung) [inline src/librustc_mir/transform/qualify_consts.rs]: Why not just `check_fn`?  That seems to be more like how the other functions here are called.

File: src/test/ui/consts/const-eval/const-eval-intrinsic-promotion.rs [hunk 7]
```diff
@@ -0,0 +1,5 @@
+#![feature(core_intrinsics)]
+fn main() {
+    let x: &'static usize =
```
Comment (RalfJung) [inline src/test/ui/consts/const-eval/const-eval-intrinsic-promotion.rs]: Please add a comment along the lines of "ensure that intrinsics do not get implicitly promoted".

Isn't there an existing test that we could reasonably add this to?

File: src/libcore/intrinsics.rs (hunk 2 above)
Comment (oli-obk) [reply src/libcore/intrinsics.rs]: Enabling the use of `const fn` for intrinsics is orthogonal and definitely on my todo list

File: src/librustc/ty/constness.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc/ty/constness.rs]: No, the function is called before we know it's an intrinsic

File: src/libcore/intrinsics.rs (hunk 2 above)
Comment (RalfJung) [reply src/libcore/intrinsics.rs]: That's fine, I am just saying that this "intermediate state" is somewhat unfortunate. But not a blocker.

This made me realize though that really we should have `#[const_stable(since)]` as well; for stable `const fn` right now there is no indication in the docs since when they are stable as `const fn`... but that's orthogonal to this PR, obviously.

File: src/librustc/ty/constness.rs (hunk 3 above)
Comment (RalfJung) [reply src/librustc/ty/constness.rs]: That seems like an odd choice IMO.

File: src/test/ui/consts/const-eval/const-eval-intrinsic-promotion.rs (hunk 7 above)
Comment (vertexclique) [reply src/test/ui/consts/const-eval/const-eval-intrinsic-promotion.rs]: There was no test for this case. This was the test that I was running locally to determine the application of whitelist.

File: src/librustc/ty/constness.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc/ty/constness.rs]: I can make it return an option. But I see this function as `is_intrinsic(did) && is_const_intrinsic(did)`

Comment (RalfJung) [reply src/librustc/ty/constness.rs]: Adding a clarifying doc comment would be enough. :)

File: src/librustc/ty/constness.rs (hunk 4 above)
Comment (oli-obk) [reply src/librustc/ty/constness.rs]: the intrinsic is named differently from the method on the integer types. I removed these duplicate entries

File: src/librustc/ty/constness.rs (hunk 5 above)
Comment (oli-obk) [reply src/librustc/ty/constness.rs]: I added them and ordered the list the same and left comments stating that these should be kept in sync.

File: src/librustc_mir/transform/qualify_consts.rs (hunk 6 above)
Comment (oli-obk) [reply src/librustc_mir/transform/qualify_consts.rs]: I have no opinion, this is what it was before I took over so I kept it. I'll rename it.

File: src/libcore/intrinsics.rs (hunk 2 above)
Comment (oli-obk) [reply src/libcore/intrinsics.rs]: I will tackle this now in a separate PR

File: src/librustc/ty/constness.rs [hunk 8]
```diff
@@ -63,13 +64,75 @@ impl<'tcx> TyCtxt<'tcx> {
...
+                    => Some(true),
+
+                    _ => Some(false)
+                }
+            }
+            _ => None
+        }
+    }
+
+    /// Checks whether the function has a `const` modifier and intrinsics can be promotable in it
```
Comment (RalfJung) [inline src/librustc/ty/constness.rs]: I don't get this doc comment. What does this have to do with promotability?

Proposal: "Checks whether the function is const-callable (ignoring any stability concerns)". Is that correct?

Comment (oli-obk) [reply src/librustc/ty/constness.rs]: Huh... idk what I was thinking

Comment (RalfJung) [reply src/librustc/ty/constness.rs]: @oli-obk reminder that this PR is almost done except for some comment fixes.

File: src/librustc_metadata/rmeta/decoder.rs [hunk 9]
```diff
@@ -1343,6 +1343,8 @@ impl<'a, 'tcx> CrateMetadata {
         let constness = match self.kind(id) {
             EntryKind::Method(data) => data.decode(self).fn_data.constness,
             EntryKind::Fn(data) => data.decode(self).constness,
+            // Some intrinsics can be const fn
+            EntryKind::ForeignFn(data) => data.decode(self).constness,
             EntryKind::Variant(..) | EntryKind::Struct(..) => hir::Constness::Const,
             _ => hir::Constness::NotConst,
         };
```
Comment (RalfJung) [inline src/librustc_metadata/rmeta/decoder.rs]: Is it expected that the logic here partially duplicates `is_const_fn_raw` in `constness.rs`?

File: src/librustc_mir/transform/qualify_min_const_fn.rs [hunk 10]
```diff
@@ -403,28 +403,28 @@ fn check_terminator(
 ///
 /// Adding more intrinsics requires sign-off from @rust-lang/lang.
 fn is_intrinsic_whitelisted(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
```
Comment (RalfJung) [inline src/librustc_mir/transform/qualify_min_const_fn.rs]: Wait, why is there *another* such whitelist now? We shouldn't have two of them. I didn't see this in earlier versions of the patch.
(Please don't amend commits during review, that makes it impossible to see what changed on your side -- because you also did a rebase, GitHub's diff is useless.)

Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: Too much stuff changed, it would have been useless to not squash. I basically removed half the implementation because it got deduplicated.

The second whitelist is for `min_const_fn`, so basically which intrinsics are reachable from stable

File: src/librustc_metadata/rmeta/decoder.rs (hunk 9 above)
Comment (oli-obk) [reply src/librustc_metadata/rmeta/decoder.rs]: kind of. This is for when the `DefId` is not local and we need to load it from metadata. Alternatively we could recompute depending on the signature instead of encoding the constness in metadata.

Comment (RalfJung) [reply src/librustc_metadata/rmeta/decoder.rs]: The implementation seems fine, but more comments would be good.

File: src/librustc_mir/transform/qualify_min_const_fn.rs (hunk 10 above)
Comment (RalfJung) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: > The second whitelist is for min_const_fn, so basically which intrinsics are reachable from stable

Any chance those could be in the same file? Or maybe even the same function, one that returns both whether the fn is const and whether it is *stable* const?

Also, if this is where stability is decided, then what is the point of the `#[rustc_const_unstable]` you added to `transmute`?

File: src/librustc_mir/transform/check_consts/validation.rs [hunk 11]
```diff
@@ -221,7 +220,7 @@ impl Validator<'a, 'mir, 'tcx> {
         let Item { tcx, body, def_id, const_kind, ..  } = *self.item;
 
         let use_min_const_fn_checks =
-            tcx.is_min_const_fn(def_id)
+            (const_kind == Some(ConstKind::ConstFn) && tcx.is_min_const_fn(def_id))
```
Comment (RalfJung) [inline src/librustc_mir/transform/check_consts/validation.rs]: Why does this (have to) change now?

Comment (oli-obk) [reply src/librustc_mir/transform/check_consts/validation.rs]: Because before I rebased, we only ever called `is_min_const_fn` on functions, but here it was called on constants and statics, too. This wasn't a problem until my changes to `is_min_const_fn`, because until my changes happened `is_min_const_fn` never called `fn_sig` on the given `DefId`, now it does. Basically `is_min_const_fn` now ICEs if you call it on a non-fn

File: src/librustc_mir/transform/qualify_min_const_fn.rs (hunk 10 above)
Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: > Also, if this is where stability is decided, then what is the point 

I think this is mainly a result of being overly scared of accidental stabilization and thus requiring two levels of activation :laughing:

I'll fix this properly with the `const_stable(since = ...)` attribute. Do you want that to happen in this PR or is it ok if I move all the lists into the same file for now?

Comment (RalfJung) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: > I think this is mainly a result of being overly scared of accidental stabilization and thus requiring two levels of activation laughing

Well, we have *four* levels of activation now, spread across 4 different files:
* the implementation
* the first whitelist in `constness.rs`
* the `rustc_const_unstable` attribute
* the second whitelist here

That's what I would call overkill. ;)

> I'll fix this properly with the const_stable(since = ...) attribute. Do you want that to happen in this PR or is it ok if I move all the lists into the same file for now?

No need to do everything in one PR. But I'd prefer to remove the `rustc_const_unstable` from this PR if that is covered by the second whitelist here, and I'd prefer the two whitelists to be merged into one *function* if possible (so there is only one list that classifies intrinsics as non-const, unstable-const, stable-const). Putting them right next to each other in one file is okay I guess if it truly is temporary.

Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: > But I'd prefer to remove the rustc_const_unstable from this PR if that is covered by the second whitelist here, 

That attribute is needed not to prevent `transmute` from being used in min const fn, but from being used in `const` items and unstable const fns.

Comment (RalfJung) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: That's exactly what I meant, this is a mess. ;)

Why does the attribute not also kick in for "min const fn"?

File: src/librustc/ty/constness.rs [hunk 12]
```diff
@@ -36,12 +36,47 @@ impl<'tcx> TyCtxt<'tcx> {
         }
     }
 
+    /// Returns `true` if the `def_id` refers to an intrisic which we've whitelisted
+    /// for being called from stable `const fn`s (`min_const_fn`).
+    ///
+    /// Adding more intrinsics requires sign-off from @rust-lang/lang.
```
Comment (RalfJung) [inline src/librustc/ty/constness.rs]: I assume the other whitelist also requires rust-lang/lang sign-off?

Please also explain the relationship between the two whitelists.

File: src/test/ui/consts/const-extern-fn/const-extern-fn-call-extern-fn.rs [hunk 13]
```diff
@@ -7,7 +7,7 @@ extern "C" {
 const extern fn bar() {
     unsafe {
         regular_in_block();
-        //~^ ERROR: cannot call functions with `"C"` abi in `min_const_fn`
+        //~^ ERROR: can only call other `const fn` within a `const fn`
```
Comment (RalfJung) [inline src/test/ui/consts/const-extern-fn/const-extern-fn-call-extern-fn.rs]: The other fn is not a `const fn` though... diagnostics regression?

Comment (bjorn3) [reply src/test/ui/consts/const-extern-fn/const-extern-fn-call-extern-fn.rs]: "can **only** call other `const fn` ..."

Comment (RalfJung) [reply src/test/ui/consts/const-extern-fn/const-extern-fn-call-extern-fn.rs]: Oh, you are right. Never mind me.^^

**Comment by RalfJung** [comment]
The code LGTM, but please add some more comments as indicated in my comments above.

File: src/librustc_mir/transform/qualify_min_const_fn.rs (hunk 10 above)
Comment (oli-obk) [reply src/librustc_mir/transform/qualify_min_const_fn.rs]: because the whitelist takes precedence over stability attributes -.-

I agree it's a mess, but if I touch it now this PR will grow a lot. It's now in a position where I can do the next PR (const_stable attribute) in a reasonable way without also doing this PR

File: src/librustc/ty/constness.rs (hunk 12 above)
Comment (oli-obk) [reply src/librustc/ty/constness.rs]: The other list is just unstable stuff, adding more unstable const stuff has so far only required lang sign off if there's no RFC justifying the experimentation

Comment (RalfJung) [reply src/librustc/ty/constness.rs]: How's that unstable, isn't that used in `const` and `static` bodies?

Comment (oli-obk) [reply src/librustc/ty/constness.rs]: ok, so the other list is the *full* list, unstable and stable, and this list is just the stable list. (The comments say so now, too)

Comment (RalfJung) [reply src/librustc/ty/constness.rs]: Oh I see, that makes sense. Thanks!

File: src/librustc/ty/constness.rs [hunk 14]
```diff
@@ -63,13 +103,81 @@ impl<'tcx> TyCtxt<'tcx> {
...
+            Abi::RustIntrinsic |
+            Abi::PlatformIntrinsic => {
+                match tcx.item_name(def_id) {
+                    // Keep this list in the same order as the match patterns in
+                    // `librustc_mir/interpret/intrinsics.rs`
+
+                    // This whitelist is a list of intrinsics that have a miri-engine implementation
+                    // and can thus be called when enabling enough feature gates. The similar
+                    // whitelist in `is_intrinsic_min_const_fn` (in this file), exists for allowing
+                    // the intrinsics to be transitively called by stable const fns.
```
Comment (RalfJung) [inline src/librustc/ty/constness.rs]: Why "transitively called"? Surely stable const fns can use a `const` item that does non-min-const-fn things?

Please also add as sentence saying that stability for other const contexts (such as const and static initializers) is controlled by the usual stability attributes at the intrinsic declaration.

File: src/librustc_metadata/rmeta/decoder.rs (hunk 9 above)
Comment (RalfJung) [reply src/librustc_metadata/rmeta/decoder.rs]: (This concern is still open.)

File: src/librustc/ty/constness.rs (hunk 14 above)
Comment (oli-obk) [reply src/librustc/ty/constness.rs]: No, stable const fns can only do min const fn things, that's the entire point of min const fn. Although we watered that down with e.g. `if` so we can make more things const fn without yet stabilizing the feature

Comment (RalfJung) [reply src/librustc/ty/constness.rs]: But I referred to a `const` *item*.

Comment (oli-obk) [reply src/librustc/ty/constness.rs]: oh yea true

File: src/librustc_metadata/rmeta/decoder.rs (hunk 9 above)
Comment (RalfJung) [reply src/librustc_metadata/rmeta/decoder.rs]: @oli-obk there's still a comment missing here.

**Comment by RalfJung** [comment]
Thanks!

@bors r+

**Comment by 9999years** [comment]
Some functions in #53718 need `intrinsics::unchecked_rem` to be made `const`; is that a candidate for the `min_const_fn` allow-list?

File: src/librustc/ty/constness.rs [hunk 15]
```diff
@@ -63,13 +103,81 @@ impl<'tcx> TyCtxt<'tcx> {
...
 
 pub fn provide(providers: &mut Providers<'_>) {
-    /// only checks whether the function has a `const` modifier
+    /// Const evaluability whitelist is here to check evaluability at the
+    /// top level beforehand.
+    fn is_const_intrinsic(tcx: TyCtxt<'_>, def_id: DefId) -> Option<bool> {
+        match tcx.fn_sig(def_id).abi() {
+            Abi::RustIntrinsic |
+            Abi::PlatformIntrinsic => {
+                match tcx.item_name(def_id) {
```
Comment (Centril) [inline src/librustc/ty/constness.rs]: Should we de-dup this list with the min-const-fn whitelist by having this one first consult the min-const-fn list?

**Comment by RalfJung** [comment]
>  Are these candidates for const intrinsics? Do we need miri-engine implementations of them?

Certainly not in this PR, this just fixes our existing checks.

File: src/librustc/ty/constness.rs (hunk 15 above)
Comment (RalfJung) [reply src/librustc/ty/constness.rs]: @oli-obk said he'd do some clean-up in a follow-up PR. For now we just want to get our const check in order.

Comment (Centril) [reply src/librustc/ty/constness.rs]: Can we leave a FIXME tho?

**Comment by oli-obk** [comment]
rebased and added FIXME

@bors r=RalfJung

**Comment by RalfJung** [comment]
Ah dang, we *finally* got Miri in shape for the next nightly...
